File: backend/app/reports/service.py
"""
Reports Module — Service

Generates formatted reports: daily, weekly, monthly, quarterly, annual,
executive summaries, owner reports, and the comprehensive portfolio report.
"""
from datetime import date, timedelta

from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.ai.service import AIAdvisorService
from app.core.config import get_settings
from app.core.database import get_db
from app.finance.service import FinancialReportingService


class ReportGenerationService:
    """Generates formatted business reports."""

    def __init__(self, session: AsyncSession):
        self.session = session
        self.finance = FinancialReportingService(session)
        self.ai = AIAdvisorService(session)

    async def generate_weekly_report(self) -> dict:
        """Generate a weekly financial summary."""
        today = date.today()
        start_of_week = today - timedelta(days=today.weekday())
        end_of_week = start_of_week + timedelta(days=6)

        summary = await self.finance.get_summary(start_of_week, end_of_week)

        return {
            "report_type": "weekly",
            "period_start": start_of_week.isoformat(),
            "period_end": end_of_week.isoformat(),
            "generated_at": today.isoformat(),
            "summary": summary.model_dump(),
        }

    async def generate_monthly_report(self, year: int, month: int) -> dict:
        """Generate a full monthly report with AI insights."""
        report = await self.finance.get_monthly_report(year, month)
        ai_analysis = await self.ai.generate_advisor_report(year)

        return {
            "report_type": "monthly",
            "year": year,
            "month": month,
            "financial_data": report.model_dump(),
            "ai_insights": ai_analysis.get("executive_summary"),
            "provider": ai_analysis.get("provider", "hostwise"),
            "recommendations": self._flatten_actions(ai_analysis)[:3],
        }

    async def generate_annual_report(self, year: int) -> dict:
        """Generate a full annual report."""
        report = await self.finance.get_annual_report(year)
        ai_analysis = await self.ai.generate_advisor_report(year)

        return {
            "report_type": "annual",
            "year": year,
            "financial_data": report.model_dump(),
            "ai_insights": ai_analysis.get("executive_summary"),
            "provider": ai_analysis.get("provider", "hostwise"),
            "recommendations": self._flatten_actions(ai_analysis),
        }

    async def generate_executive_summary(self) -> dict:
        """Executive-level summary for investors/owners."""
        today = date.today()
        current_year = today.year

        annual = await self.finance.get_annual_report(current_year)
        ai = await self.ai.generate_advisor_report(current_year)

        return {
            "report_type": "executive_summary",
            "generated_at": today.isoformat(),
            "highlights": {
                "annual_revenue": annual.summary.gross_revenue,
                "annual_profit": annual.summary.profit,
                "profit_margin": annual.summary.profit_margin,
                "yoy_growth": annual.yoy_growth,
                "property_count": annual.summary.property_count,
                "best_month": (
                    annual.best_month.model_dump() if annual.best_month else None
                ),
                "worst_month": (
                    annual.worst_month.model_dump() if annual.worst_month else None
                ),
            },
            "ai_summary": ai.get("executive_summary"),
            "provider": ai.get("provider", "hostwise"),
            "top_recommendations": [
                r for r in self._flatten_actions(ai)
                if r.get("type") in ("critical", "warning")
            ][:5],
        }

    # ── Comprehensive Portfolio Report ──────────────────────────
    # Single response powering the Reports page: executive summary,
    # AI insights, KPI comparison, property table, expense analysis,
    # risks, goals, forecast, portfolio health, and tax summary.

    @staticmethod
    def _clamp(value: float, low: float = 0.0, high: float = 100.0) -> float:
        return max(low, min(high, round(value, 1)))

    @staticmethod
    def _health_status(score: float) -> str:
        if score >= 75:
            return "excellent"
        if score >= 50:
            return "good"
        if score >= 25:
            return "average"
        return "poor"

    @staticmethod
    def _pct_change(previous: float | None, current: float) -> float | None:
        if previous is not None and previous > 0:
            return round((current - previous) / previous * 100, 1)
        return None

    @staticmethod
    def _property_card(p: dict) -> dict:
        return {
            "property_id": p.get("property_id"),
            "property_name": p.get("property_name", "Unknown"),
            "gross_revenue": round(p.get("gross_revenue", 0.0), 2),
            "net_revenue": round(p.get("net_revenue", 0.0), 2),
            "total_expenses": round(p.get("total_expenses", 0.0), 2),
            "profit": round(p.get("profit", 0.0), 2),
            "profit_margin": p.get("profit_margin", 0.0),
            "health_score": p.get("health_score"),
        }

    @staticmethod
    def _flatten_actions(ai: dict) -> list:
        """Flatten advisor `priority_actions` into a single recommendations list."""
        pa = ai.get("priority_actions") or {}
        return (
            list(pa.get("critical") or [])
            + list(pa.get("medium") or [])
            + list(pa.get("low") or [])
        )

    async def generate_portfolio_report(
        self,
        start_date: date | None = None,
        end_date: date | None = None,
        currency: str = "USD",
    ) -> dict:
        """
        Generate a comprehensive portfolio report aggregating financials,
        analytics, AI insights, risks, goals, forecast, and tax summary.

        Period-aware: pass `start_date`/`end_date` for a custom range, or omit
        them to default to the current calendar year.
        """
        from app.analytics.service import AnalyticsService
        from app.settings.service import SettingsService

        today = date.today()
        if start_date is None or end_date is None:
            start_date = start_date or date(today.year, 1, 1)
            end_date = end_date or date(today.year, 12, 31)

        # Equally-sized preceding period for comparisons
        prev_end = start_date - timedelta(days=1)
        prev_start = prev_end - (end_date - start_date)

        # Business settings affect the report (currency + tax rate)
        settings = await SettingsService(self.session).get_all()
        if not currency:
            currency = settings.get("default_currency", "USD")
        tax_rate = float(settings.get("tax_rate", 0) or 0)

        # Financial data (current + previous period for comparisons)
        annual = await self.finance.get_period_report(start_date, end_date)
        prev_annual = await self.finance.get_period_report(prev_start, prev_end)

        # Portfolio analytics (property ranking, health, forecast)
        analytics = AnalyticsService(self.session)
        portfolio = await analytics.get_portfolio_analytics(
            start_date=start_date, end_date=end_date
        )

        # AI insights & recommendations (rules by default, LLM when configured)
        ai = await self.ai.generate_advisor_report(
            start_date=start_date, end_date=end_date
        )

        # ── Property performance (enriched with health) ──
        ranking_map = {
            str(r["property_id"]): r for r in portfolio.get("property_ranking", [])
        }
        property_performance = []
        for p in annual.revenue_by_property:
            rank = ranking_map.get(str(p.property_id), {})
            property_performance.append({
                **p.model_dump(),
                "health_score": rank.get("health_score"),
            })
        property_performance.sort(key=lambda x: x["net_revenue"], reverse=True)

        best = property_performance[0] if property_performance else None
        worst = property_performance[-1] if property_performance else None

        # ── KPI comparison (previous vs current) ──
        kpi_comparison = {
            "revenue": {
                "previous": round(prev_annual.summary.net_revenue, 2),
                "current": round(annual.summary.net_revenue, 2),
                "change_pct": self._pct_change(
                    prev_annual.summary.net_revenue, annual.summary.net_revenue
                ),
            },
            "profit": {
                "previous": round(prev_annual.summary.profit, 2),
                "current": round(annual.summary.profit, 2),
                "change_pct": self._pct_change(
                    prev_annual.summary.profit, annual.summary.profit
                ),
            },
            "expenses": {
                "previous": round(prev_annual.summary.total_expenses, 2),
                "current": round(annual.summary.total_expenses, 2),
                "change_pct": self._pct_change(
                    prev_annual.summary.total_expenses, annual.summary.total_expenses
                ),
            },
        }

        # ── Portfolio health ──
        portfolio_health = self._compute_portfolio_health(
            portfolio, annual, kpi_comparison
        )

        # ── Executive summary ──
        executive_summary = {
            "period_start": start_date.isoformat(),
            "period_end": end_date.isoformat(),
            "gross_revenue": annual.summary.gross_revenue,
            "net_profit": annual.summary.profit,
            "profit_margin": annual.summary.profit_margin,
            "property_count": annual.summary.property_count,
            "best_property": self._property_card(best) if best else None,
            "worst_property": self._property_card(worst) if worst else None,
            "portfolio_health_score": portfolio_health["score"],
            "portfolio_health_status": portfolio_health["status"],
        }

        # ── Derived sections ──
        ai_insights = self._build_ai_insights(ai, kpi_comparison, annual, portfolio)
        expense_analysis = self._build_expense_analysis(annual, prev_annual)
        risks = self._build_risks(ai, kpi_comparison, annual, best)
        goals = await self._build_goals(prev_annual, portfolio)
        forecast = self._build_forecast(start_date, end_date, portfolio, annual)

        taxable_income = round(
            annual.summary.net_revenue - annual.summary.total_expenses, 2
        )
        tax_summary = {
            "rental_income": round(annual.summary.net_revenue, 2),
            "deductible_expenses": round(annual.summary.total_expenses, 2),
            "estimated_taxable_income": taxable_income,
            "tax_rate": tax_rate,
            "estimated_tax_liability": round(taxable_income * tax_rate / 100, 2),
        }

        return {
            "report_type": "portfolio",
            "year": end_date.year,
            "provider": ai.get("provider", "hostwise"),
            "generated_at": today.isoformat(),
            "organization": get_settings().APP_NAME,
            "currency": currency.upper(),
            "period_start": start_date.isoformat(),
            "period_end": end_date.isoformat(),
            "period": {
                "start": start_date.isoformat(),
                "end": end_date.isoformat(),
                "label": self._format_period_label(start_date, end_date),
                "days": (end_date - start_date).days + 1,
            },
            "previous_period": {
                "start": prev_start.isoformat(),
                "end": prev_end.isoformat(),
                "label": self._format_period_label(prev_start, prev_end),
                "days": (prev_end - prev_start).days + 1,
            },
            "executive_summary": executive_summary,
            "ai_insights": ai_insights,
            "kpi_comparison": kpi_comparison,
            "property_performance": property_performance,
            "monthly_breakdown": [m.model_dump() for m in annual.monthly_breakdown],
            "expense_analysis": expense_analysis,
            "best_worst_properties": {
                "best": self._property_card(best) if best else None,
                "worst": self._property_card(worst) if worst else None,
            },
            "risks": risks,
            "goals": goals,
            "forecast": forecast,
            "portfolio_health": portfolio_health,
            "tax_summary": tax_summary,
        }

    def _compute_portfolio_health(
        self,
        portfolio: dict,
        annual,
        kpi: dict,
    ) -> dict:
        """Compute a portfolio health score (0-100) + component bars."""
        # Component scores (used for the visual bars)
        rev_growth = kpi["revenue"]["change_pct"]
        revenue_score = self._clamp(50 + (rev_growth or 0) * 2.5)

        margin = annual.summary.profit_margin
        profit_score = self._clamp(margin / 50 * 100)

        net = annual.summary.net_revenue
        expense_ratio = (annual.summary.total_expenses / net * 100) if net > 0 else 0.0
        expense_score = self._clamp(100 - expense_ratio * 0.8)

        # Overall score: average property health when available,
        # otherwise a weighted blend of the component scores.
        ranking = portfolio.get("property_ranking", [])
        if ranking:
            score = round(
                sum(r.get("health_score", 50) for r in ranking) / len(ranking), 1
            )
        else:
            score = round(
                revenue_score * 0.30
                + profit_score * 0.35
                + expense_score * 0.35,
                1,
            )

        return {
            "score": score,
            "status": self._health_status(score),
            "components": {
                "revenue": revenue_score,
                "profit": profit_score,
                "expenses": expense_score,
            },
            "distribution": portfolio.get(
                "health_distribution",
                {"excellent": 0, "good": 0, "average": 0, "poor": 0},
            ),
        }

    def _build_ai_insights(
        self,
        ai: dict,
        kpi: dict,
        annual,
        portfolio: dict,
    ) -> dict:
        """Natural-language insights + drivers + top risk/recommendation."""
        drivers = []

        exp = kpi["expenses"]["change_pct"]
        if exp is not None:
            direction = "down" if exp <= 0 else "up"
            drivers.append({
                "label": "Total expenses",
                "detail": f"{direction.capitalize()} {abs(exp):.1f}%",
            })

        for c in annual.expense_by_category:
            if "maintenance" in c.category_name.lower() or "repair" in c.category_name.lower():
                drivers.append({
                    "label": "Maintenance costs",
                    "detail": f"{c.total:,.0f} ({c.percentage}% of expenses)",
                })
                break

        recs = self._flatten_actions(ai)
        critical = [r for r in recs if r.get("type") == "critical"]
        warnings = [r for r in recs if r.get("type") == "warning"]
        top = (critical or warnings or recs or [None])[0]

        biggest_risk = None
        if top:
            biggest_risk = {
                "level": top.get("type"),
                "title": top.get("title"),
                "cause": top.get("cause"),
                "suggested_action": top.get("suggested_action"),
            }

        recommendation = (
            biggest_risk["suggested_action"]
            if biggest_risk and biggest_risk.get("suggested_action")
            else "No critical actions required — the portfolio is on track."
        )

        return {
            "summary": ai.get("executive_summary", ""),
            "provider": ai.get("provider", "hostwise"),
            "revenue_change_pct": portfolio.get("revenue_growth_yoy"),
            "drivers": drivers,
            "biggest_risk": biggest_risk,
            "recommendation": recommendation,
            "recommendations": recs[:5],
        }

    def _build_expense_analysis(self, annual, prev_annual) -> dict:
        """Expense categories with biggest / smallest / fastest-growing."""
        categories = [c.model_dump() for c in annual.expense_by_category]
        prev_map = {c.category_name: c.total for c in prev_annual.expense_by_category}

        for c in categories:
            prev_total = prev_map.get(c["category_name"], 0.0)
            c["growth_pct"] = (
                round((c["total"] - prev_total) / prev_total * 100, 1)
                if prev_total > 0 else None
            )

        if not categories:
            return {"categories": [], "biggest": None, "smallest": None, "fastest_growing": None}

        biggest = max(categories, key=lambda x: x["total"])
        smallest = min(categories, key=lambda x: x["total"])
        growing = [
            c for c in categories
            if c.get("growth_pct") is not None and c["total"] > 0
        ]
        fastest_growing = max(growing, key=lambda x: x["growth_pct"]) if growing else None

        return {
            "categories": categories,
            "biggest": biggest,
            "smallest": smallest,
            "fastest_growing": fastest_growing,
        }

    def _build_risks(
        self,
        ai: dict,
        kpi: dict,
        annual,
        best: dict | None,
    ) -> list[dict]:
        """Business risks: AI recommendations + computed risk signals."""
        risks: list[dict] = []
        seen: set[str] = set()

        def add(level: str, title: str, detail: str) -> None:
            key = title.lower()
            if key in seen:
                return
            seen.add(key)
            risks.append({"level": level, "title": title, "detail": detail})

        for r in self._flatten_actions(ai):
            if r.get("type") in ("critical", "warning"):
                add(
                    "high" if r["type"] == "critical" else "medium",
                    r.get("title", "Risk"),
                    r.get("cause", ""),
                )

        if best and annual.summary.net_revenue > 0:
            share = best["net_revenue"] / annual.summary.net_revenue * 100
            if share > 50:
                add(
                    "medium",
                    "Revenue concentration",
                    f"{best['property_name']} generates {share:.0f}% of total "
                    f"revenue — heavy reliance on a single property.",
                )

        net_by_month = [
            m.net_revenue for m in annual.monthly_breakdown if m.net_revenue > 0
        ]
        if len(net_by_month) >= 4:
            peak, trough = max(net_by_month), min(net_by_month)
            if peak / trough > 3:
                add(
                    "medium",
                    "Seasonal dependency",
                    f"Revenue swings from {trough:,.0f} to {peak:,.0f} per month — "
                    f"a {peak / trough:.1f}x seasonal gap.",
                )

        if annual.summary.net_revenue > 0:
            ratio = annual.summary.total_expenses / annual.summary.net_revenue * 100
            if ratio > 60:
                add(
                    "medium",
                    "High expense ratio",
                    f"Expenses consume {ratio:.0f}% of net revenue, squeezing profitability.",
                )

        return risks

    async def _build_goals(self, prev_annual, portfolio: dict) -> dict:
        """Revenue goals with current progress."""
        from app.properties.repository import PropertyRepository

        prop_repo = PropertyRepository(self.session)
        props = await prop_repo.get_all_properties()

        revenue_goal = sum(
            p.target_annual_revenue for p in props if p.target_annual_revenue
        )
        if not revenue_goal:
            revenue_goal = (
                round(prev_annual.summary.net_revenue * 1.2, 2)
                if prev_annual.summary.net_revenue > 0 else 0.0
            )

        current_net = portfolio.get("net_revenue", 0.0)
        revenue_progress = (
            round(current_net / revenue_goal * 100, 1) if revenue_goal > 0 else 0.0
        )

        return {
            "revenue": {
                "goal": round(revenue_goal, 2),
                "current": round(current_net, 2),
                "progress": self._clamp(revenue_progress),
            },
        }

    def _build_forecast(
        self,
        start_date,
        end_date,
        portfolio: dict,
        annual,
    ) -> dict:
        """Next-quarter revenue forecast with confidence."""
        today = date.today()

        months = [
            m for m in annual.monthly_breakdown
            if m.gross_revenue > 0 or m.net_revenue > 0
        ]
        # Only count completed months when the range reaches the present
        if end_date >= today and months:
            months = [
                m for m in months
                if m.year < today.year or (m.year == today.year and m.month <= today.month)
            ]

        relevant = months[-3:] if months else []
        pool = relevant or months

        if pool:
            avg_monthly = sum(m.net_revenue for m in pool) / len(pool)
        else:
            avg_monthly = portfolio.get("net_revenue", 0.0) / max(len(annual.monthly_breakdown), 1)

        data_months = len(months)
        confidence = round(
            min(95, 45 + data_months * 3 + (12 if data_months >= 12 else 0))
        )

        return {
            "next_quarter_revenue": round(avg_monthly * 3, 2),
            "confidence": confidence,
        }

    @staticmethod
    def _format_period_label(start: date, end: date) -> str:
        return f"{start.strftime('%b %d, %Y')} – {end.strftime('%b %d, %Y')}"


async def get_report_service(
    session: AsyncSession = Depends(get_db),
) -> ReportGenerationService:
    return ReportGenerationService(session)

File: backend-rs/src/reports/service.rs
//! Reports service: generates weekly/monthly/annual/executive/portfolio report
//! JSON, reusing the finance + analytics + AI services. Mirrors
//! `backend/app/reports/service.py`.

use chrono::{Datelike, Duration as ChronoDuration, Local};
use serde_json::{json, Value};
use sqlx::SqlitePool;

use crate::ai::service::{flatten_actions, AiAdvisorService};
use crate::analytics::service as analytics;
use crate::core::error::AppError;
use crate::finance::service::FinanceService;
use crate::settings::service as settings;

fn round2(v: f64) -> f64 {
    (v * 100.0).round() / 100.0
}

/// Shift an inclusive [start, end] window one year back (YoY comparison).
fn shift_window_back(start: &str, end: &str) -> (String, String) {
    let parse = |s: &str| chrono::NaiveDate::parse_from_str(s, "%Y-%m-%d").ok();
    match (parse(start), parse(end)) {
        (Some(sd), Some(ed)) => {
            let ps = sd.checked_sub_months(chrono::Months::new(12)).unwrap_or(sd);
            let pe = ed.checked_sub_months(chrono::Months::new(12)).unwrap_or(ed);
            (
                ps.format("%Y-%m-%d").to_string(),
                pe.format("%Y-%m-%d").to_string(),
            )
        }
        _ => {
            let y = chrono::Local::now().year() - 1;
            (format!("{y:04}-01-01"), format!("{y:04}-12-31"))
        }
    }
}

pub struct ReportGenerationService {
    pub pool: SqlitePool,
}

impl ReportGenerationService {
    pub fn from_pool(pool: SqlitePool) -> Self {
        Self { pool }
    }

    fn advisor(&self) -> AiAdvisorService {
        AiAdvisorService {
            pool: self.pool.clone(),
        }
    }

    pub async fn generate_weekly_report(&self) -> Result<Value, AppError> {
        let today = Local::now().date_naive();
        let start = today - ChronoDuration::days(today.weekday().num_days_from_monday() as i64);
        let end = start + ChronoDuration::days(6);
        let s = start.format("%Y-%m-%d").to_string();
        let e = end.format("%Y-%m-%d").to_string();
        let summary = FinanceService::from_pool(self.pool.clone())
            .get_summary(Some(&s), Some(&e))
            .await?;
        Ok(json!({
            "report_type": "weekly",
            "period_start": s,
            "period_end": e,
            "generated_at": today.format("%Y-%m-%d").to_string(),
            "summary": serde_json::to_value(&summary)?,
        }))
    }

    pub async fn generate_monthly_report(&self, year: i32, month: i64) -> Result<Value, AppError> {
        let finance = FinanceService::from_pool(self.pool.clone());
        let report = finance.get_monthly_report(year, month as u32).await?;
        let ai = self
            .advisor()
            .generate_advisor_report(Some(year), None, None)
            .await?;
        let recs: Vec<Value> = flatten_actions(&ai).into_iter().take(3).collect();
        Ok(json!({
            "report_type": "monthly",
            "year": year,
            "month": month,
            "financial_data": serde_json::to_value(&report)?,
            "ai_insights": ai["executive_summary"],
            "provider": ai["provider"],
            "recommendations": recs,
        }))
    }

    pub async fn generate_annual_report(&self, year: i32) -> Result<Value, AppError> {
        let finance = FinanceService::from_pool(self.pool.clone());
        let report = finance.get_annual_report(year).await?;
        let ai = self
            .advisor()
            .generate_advisor_report(Some(year), None, None)
            .await?;
        Ok(json!({
            "report_type": "annual",
            "year": year,
            "financial_data": serde_json::to_value(&report)?,
            "ai_insights": ai["executive_summary"],
            "provider": ai["provider"],
            "recommendations": flatten_actions(&ai),
        }))
    }

    pub async fn generate_executive_summary(&self) -> Result<Value, AppError> {
        let today = Local::now().date_naive();
        let year = today.year();
        let finance = FinanceService::from_pool(self.pool.clone());
        let annual = finance.get_annual_report(year).await?;
        let ai = self
            .advisor()
            .generate_advisor_report(Some(year), None, None)
            .await?;

        let property_count: i64 =
            sqlx::query_scalar("SELECT COUNT(*) FROM properties WHERE is_deleted = 0")
                .fetch_one(&self.pool)
                .await?;
        let profit_margin = if annual.summary.net_revenue > 0.0 {
            (annual.summary.profit / annual.summary.net_revenue) * 100.0
        } else {
            0.0
        };
        let best = annual
            .monthly_breakdown
            .iter()
            .max_by(|a, b| a.profit.total_cmp(&b.profit));
        let worst = annual
            .monthly_breakdown
            .iter()
            .min_by(|a, b| a.profit.total_cmp(&b.profit));

        let recs = flatten_actions(&ai);
        let top: Vec<Value> = recs
            .into_iter()
            .filter(|r| matches!(r["type"].as_str(), Some("critical") | Some("warning")))
            .take(5)
            .collect();

        Ok(json!({
            "report_type": "executive_summary",
            "generated_at": today.format("%Y-%m-%d").to_string(),
            "highlights": {
                "annual_revenue": annual.summary.net_revenue,
                "annual_expenses": annual.summary.total_expenses,
                "annual_profit": annual.summary.profit,
                "profit_margin": profit_margin,
                "property_count": property_count,
                "best_month": best.map(|m| serde_json::to_value(m).unwrap_or(Value::Null)),
                "worst_month": worst.map(|m| serde_json::to_value(m).unwrap_or(Value::Null)),
            },
            "ai_summary": ai["executive_summary"],
            "provider": ai["provider"],
            "top_recommendations": top,
        }))
    }

    pub async fn generate_portfolio_report(
        &self,
        start: Option<&str>,
        end: Option<&str>,
        currency: Option<&str>,
    ) -> Result<Value, AppError> {
        let today = Local::now().date_naive();
        let (s, e) = match (start, end) {
            (Some(a), Some(b)) => (a.to_string(), b.to_string()),
            _ => {
                let y = today.year();
                (format!("{y:04}-01-01"), format!("{y:04}-12-31"))
            }
        };

        let all = settings::get_all(&self.pool).await?;
        let cur = currency
            .filter(|c| c.len() == 3)
            .map(|c| c.to_string())
            .unwrap_or_else(|| {
                all["default_currency"]
                    .as_str()
                    .unwrap_or("EUR")
                    .to_string()
            });
        let tax_rate = all["tax_rate"].as_f64().unwrap_or(0.0);

        let finance = FinanceService::from_pool(self.pool.clone());
        let summary = finance.get_summary(Some(&s), Some(&e)).await?;
        let portfolio =
            analytics::get_portfolio_analytics(&self.pool, None, Some(&s), Some(&e)).await?;
        let ai = self
            .advisor()
            .generate_advisor_report(None, Some(&s), Some(&e))
            .await?;

        let net = summary.net_revenue;
        let expenses = summary.total_expenses;
        let profit = net - expenses;
        let profit_margin = if net > 0.0 { profit / net * 100.0 } else { 0.0 };

        let year = s[0..4].parse::<i32>().unwrap_or_else(|_| today.year());
        let annual = finance.get_annual_report(year).await?;
        let (prev_s, prev_e) = shift_window_back(&s, &e);
        let prev_summary = finance.get_summary(Some(&prev_s), Some(&prev_e)).await?;

        let pct_change = |prev: f64, cur: f64| -> Option<f64> {
            if prev > 0.0 {
                Some(((cur - prev) / prev) * 100.0)
            } else {
                None
            }
        };
        let kpi_comparison = json!({
            "revenue": {
                "previous": prev_summary.net_revenue,
                "current": net,
                "change_pct": pct_change(prev_summary.net_revenue, net),
            },
            "profit": {
                "previous": prev_summary.profit,
                "current": profit,
                "change_pct": pct_change(prev_summary.profit, profit),
            },
            "expenses": {
                "previous": prev_summary.total_expenses,
                "current": expenses,
                "change_pct": pct_change(prev_summary.total_expenses, expenses),
            },
        });

        let property_performance: Vec<Value> = portfolio["properties"]
            .as_array()
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .map(|p| {
                json!({
                    "property_id": p["property_id"],
                    "property_name": p["name"],
                    "net_revenue": p["net_revenue"],
                    "total_expenses": p["expenses"],
                    "profit": p["profit"],
                    "profit_margin": p["profit_margin"],
                    "reservation_count": p["reservations"],
                })
            })
            .collect();

        let exp_categories = portfolio["expense_categories"]
            .as_array()
            .cloned()
            .unwrap_or_default();
        let expense_analysis = json!({
            "categories": exp_categories,
            "biggest": exp_categories.first().cloned(),
            "smallest": exp_categories.last().cloned(),
            "fastest_growing": null,
        });

        let ranking = portfolio["property_ranking"]
            .as_array()
            .cloned()
            .unwrap_or_default();
        let best_worst_properties = json!({
            "best": ranking.first().cloned(),
            "worst": ranking.last().cloned(),
        });

        let health = ai["health_score"].clone();
        let portfolio_health = json!({
            "score": health["score"],
            "status": health["status"],
            "components": health["components"],
            "distribution": portfolio["health_distribution"],
        });

        let forecast_next = portfolio["forecast_next_month"].as_f64().unwrap_or(0.0);
        let forecast = json!({
            "next_quarter_revenue": round2(forecast_next * 3.0),
            "confidence": 0.6,
        });

        let revenue_goal = all["annual_revenue_goal"].as_f64().unwrap_or(0.0);
        let goals = json!({
            "revenue": {
                "goal": revenue_goal,
                "current": net,
                "progress": if revenue_goal > 0.0 {
                    round2(net / revenue_goal * 100.0)
                } else {
                    0.0
                },
            }
        });

        let tax_summary = json!({
            "rental_income": portfolio["total_gross_revenue"],
            "deductible_expenses": expenses,
            "estimated_taxable_income": profit.max(0.0),
            "tax_rate": tax_rate,
            "estimated_tax": (profit * tax_rate / 100.0).max(0.0),
        });

        let org = all["organization_name"]
            .as_str()
            .unwrap_or("HostWise")
            .to_string();

        Ok(json!({
            "report_type": "portfolio",
            "year": year,
            "period": format!("{s} to {e}"),
            "previous_period": format!("{prev_s} to {prev_e}"),
            "period_start": s,
            "period_end": e,
            "generated_at": today.format("%Y-%m-%d").to_string(),
            "organization": org,
            "currency": cur,
            "executive_summary": {
                "summary": ai["executive_summary"],
                "provider": ai["provider"],
                "health_status": health["status"],
            },
            "ai_insights": {
                "recommendations": flatten_actions(&ai),
                "opportunities": ai["opportunities"],
                "risks": ai["risks"],
                "forecast": forecast,
            },
            "kpi_comparison": kpi_comparison,
            "property_performance": property_performance,
            "monthly_breakdown": serde_json::to_value(&annual.monthly_breakdown)?,
            "expense_analysis": expense_analysis,
            "best_worst_properties": best_worst_properties,
            "risks": ai["risks"],
            "opportunities": ai["opportunities"],
            "goals": goals,
            "forecast": forecast,
            "portfolio_health": portfolio_health,
            "tax_summary": tax_summary,
            // Legacy keys used by older consumers.
            "provider": ai["provider"],
            "health": health,
            "kpis": {
                "gross_revenue": portfolio["total_gross_revenue"],
                "net_revenue": net,
                "total_expenses": expenses,
                "profit": profit,
                "profit_margin": profit_margin,
                "reservation_count": portfolio["reservation_count"],
                "nights": portfolio["nights"],
                "cancellation_rate": portfolio["cancellation_rate"],
                "revenue_count": summary.revenue_count,
                "expense_count": summary.expense_count,
            },
            "property_table": portfolio["properties"],
            "recommendations": flatten_actions(&ai),
        }))
    }
}
